sycl/include/sycl/ext/oneapi/weak_object_base.hpp

```diff
@@ -0,0 +1,68 @@
+//==------- weak_object_base.hpp --- SYCL weak objects base class ----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/detail/defines_elementary.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace detail {
+template <typename SYCLObjT> class weak_object_base;
+
+// Helper function for getting the underlying weak_ptr from a weak_object.
+template <typename SYCLObjT>
+decltype(weak_object_base<SYCLObjT>::MObjWeakPtr)
+getSyclWeakObjImpl(const weak_object_base<SYCLObjT> &WeakObj) {
+  return WeakObj.MObjWeakPtr;
+}
+
+// Common base class for weak_object.
+template <typename SYCLObjT> class weak_object_base {
+public:
+  using object_type = SYCLObjT;
+
+  constexpr weak_object_base() noexcept : MObjWeakPtr() {}
+  weak_object_base(const SYCLObjT &SYCLObj) noexcept
+      : MObjWeakPtr(sycl::detail::getSyclObjImpl(SYCLObj)) {}
+  weak_object_base(const weak_object_base &Other) noexcept
+      : MObjWeakPtr(Other.MObjWeakPtr) {}
+  weak_object_base(weak_object_base &&Other) noexcept
+      : MObjWeakPtr(std::move(Other.MObjWeakPtr)) {}
```

Comment:
My C++ kung fu isn't strong enough to really know, but do we need those explicit versions? Are implicit ones disabled for some reasons? Even if they are, would ` = default;` work?

Reply:
I don't think there's anything stopping us from using the implicit ones, I just don't trust them and they're easy to break. I have changed them to `= default`.